aws_lambda/policy_checks.py

```diff
@@ -0,0 +1,397 @@
+        unprotected_branches = []
+
+        for branch in json:
+            if branch["protected"] == False:
+                branches_unprotected = True
+
+                unprotected_branches.append(branch["name"])
+
+        return branches_unprotected
+
+    else:
+        return f"Error {branches_response.status_code}: {branches_response.json()["message"]}"
+
+# Repos with unsigned commits
+def check_signed_commits(repo_api_url: str, gh: api_controller) -> bool | str:
+    """
+    Checks if the last 15 commits are signed.
+    If any commit is not signed, return True
+
+    Args:
+        repo_api_url (str): The API endpoint to get the repository's commits
+        gh (api_controller): An instance of the api_controller class to make calls to the GitHub API
+    Returns:
+        bool (True if check breaks policy)
+        or
+        str (an error has occured when accessing the API)
+    """
+
+    commits_response = gh.get(repo_api_url, {"per_page": 15}, False)
+
+    if commits_response.status_code == 200:
+        commits = commits_response.json()
+
+        unsigned_commits = False
+
+        for commit in commits:
+            if commit["commit"]["verification"]["verified"] == False:
+                unsigned_commits = True
+
+        return unsigned_commits
+    elif commits_response.status_code == 409:
+        # If error 409, then the repository does not have any commits.
+        return False
+    else:
+        return f"Error {commits_response.status_code}: {commits_response.json()["message"]}"
+
+# Repos without README.md, Liscense file (public only), 
+# PIRR.md (private or internal) and .gitignore
+
+def check_file_exists(repo_api_url: str, gh: api_controller, files: list[str]):
+    """
+    Checks if a given filename exists in the repository.
+    Returns True if file is missing
+
+    Args:
+        repo_api_url (str): The API endpoint to get the repository's contents
+        gh (api_controller): An instance of the api_controller class to make calls to the GitHub API
+        files (list[str]): The list of filenames to look for
+    Returns:
+        bool (True if check breaks policy)
+    """
+
+    for file in files:
+        if gh.get(repo_api_url.replace("{+path}", file), {}, False).status_code == 200:
+            return False
+        
+    return True
+
+
+# Any external PR's
+def check_external_pr(repo_api_url: str, repo_full_name: str, gh: api_controller) -> bool | str:
+    """
+    Checks if there are any pull requests in the repository from non-members of the organisation.
+    Returns True if there is an external pull request
+
+    Args:
+        repo_api_url (str): The API endpoint to get the repository's pull requests
+        repo_full_name (str): The fullname of the repository from the GitHub API (format: <owner>/<repo>)
+        gh (api_controller): An instance of the api_controller class to make calls to the GitHub API
+    Returns:
+        bool (True if check breaks policy)
```

Comment:
Is this actually breaking policy or are we just flagging repos where externals have made PR requests?

Reply:
This was a requested feature by security. It's also suggested by clause 5.5 in the GitHub Usage Policy